orttraining/orttraining/core/framework/gradient_graph_builder.cc

```diff
@@ -255,21 +256,84 @@ Status GradientGraphBuilder::Build(const std::unordered_set<std::string>* p_init
 
   // so far, visited are the minimum node in between
   // visited_node_args are the node_args involved
+  std::unordered_map<std::string, int> input_arg_ref_count;
+  for (auto node : visited) {
+    if (node->OpType() == "LayerNormalization") {
+      continue;
+    }
+
+    //TODO: might not need two sets, the union of them might be enough
+    std::unordered_set<std::string> input_args_need_grad, output_args_need_grad;
+    for (auto arg : node->InputDefs()) {
+      if (visited_node_args.find(arg) != visited_node_args.end()) {
+        input_args_need_grad.insert(arg->Name());
+      }
+
+      if (input_arg_ref_count.find(arg->Name()) == input_arg_ref_count.end()) {
+        input_arg_ref_count[arg->Name()] = 1;
+      } else {
+        input_arg_ref_count[arg->Name()] += 1;
+      }
+    }
+    for (auto arg : node->OutputDefs()) {
+      if (visited_node_args.find(arg) != visited_node_args.end()) {
+        output_args_need_grad.insert(arg->Name());
+      }
+    }
+
+    GradientDef node_defs = GetGradientForOp(gradient_graph_config_, graph_, node, output_args_need_grad,
+                                             input_args_need_grad, logger_, input_arg_ref_count);
+
+    if (node_defs.empty()) {
+      LOGS(logger_, WARNING) << "GetGradientForOp() did not create any nodes for node "
+                             << node->Name() << " of type " << node->OpType() << ".";
+    }
+
+    // updates arg name if gradient accumulation is needed
+    for (auto& op_def : node_defs) {
+      for (auto& arg : op_def.output_args) {
+        auto found = pending_.find(arg.name);
+        if (found != pending_.end() && found->second > 1) {
+          auto idx = gradients_to_accumulate_[arg].size();
+          std::string indexed_arg_name = arg.name + "_" + to_string(idx);
+          gradients_to_accumulate_[arg].push_back(ArgDef(indexed_arg_name, arg.type_proto));
+
+          arg.name = indexed_arg_name;
+        }
+      }
+    }
+    gradient_graph_defs.AddNodeDefs(node_defs);
+  }
+
+  // Run a second pass on nodes that have invertible ops.
   for (auto node : visited) {
+    if (node->OpType() != "LayerNormalization") {
+      continue;
+    }
+
     //TODO: might not need two sets, the union of them might be enough
     std::unordered_set<std::string> input_args_need_grad, output_args_need_grad;
     for (auto arg : node->InputDefs()) {
       if (visited_node_args.find(arg) != visited_node_args.end()) {
         input_args_need_grad.insert(arg->Name());
       }
+      if (input_arg_ref_count.find(arg->Name()) == input_arg_ref_count.end()) {
+        input_arg_ref_count[arg->Name()] = 1;
+      } else {
+        input_arg_ref_count[arg->Name()] += 1;
+      }
```

Comment:
This is the ref count in the forward graph. They are not the ones got stashed. 

Instead, we should examie the nodes in node_defs, retunred by GetGradientForOp().

Reply:
I'll sync offline on this.